File: graph.py
import networkx as nx
import matplotlib.pyplot as plt

# Step 1: Load the Zachary's Karate Club graph
G = nx.karate_club_graph()

# Step 2: Compute the force-directed layout (spring layout)
pos = nx.spring_layout(G, seed=42)  # seed for reproducibility

# Step 3: Set up the plot
plt.figure(figsize=(10, 8))
plt.title("Zachary's Karate Club Graph (Force-Directed Layout)")

# Step 4: Draw nodes
# We color the nodes by a node attribute -  "club" affiliation 
club_colors = {
    'Mr. Hi': 'skyblue',
    'Officer': 'lightgreen'
}
node_colors = [club_colors[G.nodes[n]['club']] for n in G.nodes]

nx.draw_networkx_nodes(G, pos, node_color=node_colors, node_size=500)

# Step 5: Draw edges
nx.draw_networkx_edges(G, pos, alpha=0.6)

# Step 6: Draw node labels
nx.draw_networkx_labels(G, pos, font_size=10, font_color='black')

# Step 7: Remove axes
plt.axis('off')

# Step 8: Show the plot
plt.show()

File: graph.R
# Step 1: Load the igraph package
library(igraph)

# Step 2: Load the built-in Karate Club graph
g <- make_graph("Zachary")  # Zachary's Karate Club graph

# Step 3: Optionally inspect the graph
# print(g)
# summary(g)

# Step 4: Compute a force-directed layout (Fruchterman-Reingold algorithm)
layout <- layout_with_fr(g)

# Step 5: Plot the graph using the layout
plot(
  g,
  layout = layout,         # Use computed layout
  vertex.size = 20,        # Size of the nodes
  vertex.label = V(g)$name,# Node labels
  vertex.color = "skyblue",# Node color
  edge.arrow.size = 0.5,   # Arrow size for edges (not really used here, it's undirected)
  main = "Karate Club Graph (Force-Directed Layout)"
)
